third_party/nginx_latency_status_module/src/ngx_http_latency_stub_status_module.c

```diff
@@ -0,0 +1,306 @@
+                       "invalid value for latency_status_sub: scaled_factor %d. It must be greater than 0", scale_factor);
+    return NGX_CONF_ERROR;
+  }
+  if (max_exponent < 1) {
+    ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                       "invalid value for latency_status_sub: max_exponent %d. It must be at least 1", max_exponent);
+    return NGX_CONF_ERROR;
+  }
+
+  main_conf->base = base;
+  main_conf->scale_factor = scale_factor;
+  main_conf->max_exponent = max_exponent;
+  shm_max_exponent = max_exponent;
+
+  ngx_int_t* bucket_bounds;
+  ngx_int_t power = 1;
+  bucket_bounds = ngx_palloc(cf->pool, sizeof(ngx_int_t) * (max_exponent + 1));
+  if (bucket_bounds == NULL) {
+    return NGX_CONF_ERROR;
+  }
+  for (ngx_int_t i = 0; i <= max_exponent; i++) {
+    bucket_bounds[i] = scale_factor * power;
+    power *= base;
+  }
+  main_conf->latency_bucket_bounds = bucket_bounds;
+
+  ngx_shm_zone_t *shm_zone;
+  ngx_str_t *shm_name;
+
+  shm_name = ngx_palloc(cf->pool, sizeof *shm_name);
+  shm_name->len = sizeof("latency_shared_memory") - 1;
+  shm_name->data = (unsigned char *) "shared_memory";
+  shm_zone = ngx_shared_memory_add(cf, shm_name, 8 * ngx_pagesize, &ngx_http_latency_stub_status_module);
+
+  if(shm_zone == NULL) {
+    return NGX_CONF_ERROR;
+  }
+
+  shm_zone->init = ngx_http_latency_init_shm_zone;
+  main_conf->shm_zone = shm_zone;
+
+  main_conf->enabled = 1;
+  return NGX_CONF_OK;
+}
+
+
+// Create a new ngx_http_latency_main_conf_t.
+void *ngx_http_latency_create_main_conf(ngx_conf_t* cf){
+  ngx_http_latency_main_conf_t* conf;
+
+  conf = ngx_pcalloc(cf->pool, sizeof(ngx_http_latency_main_conf_t));
+  if(conf == NULL) {
+    return NGX_CONF_ERROR;
+  }
+
+  conf->enabled = NGX_CONF_UNSET;
+  return conf;
+}
+
+
+// Create a new ngx_http_latency_conf_t.
+void *ngx_http_latency_create_loc_conf(ngx_conf_t *cf)
+{
+  ngx_http_latency_conf_t *conf;
+
+  conf = ngx_pcalloc(cf->pool, sizeof(ngx_http_latency_conf_t));
+  if (conf == NULL) {
+    return NULL;
+  }
+
+  conf->receiver_enabled = NGX_CONF_UNSET;
+  conf->status_page_enabled = NGX_CONF_UNSET;
+  return conf;
+}
+
+
+// Merge ngx_http_latency_conf_t configs in case they have been set in multiple
+// locations.
+char *ngx_http_latency_merge_loc_conf(ngx_conf_t *cf, void *parent, void *child)
+{
+  ngx_http_latency_conf_t *prev = parent;
+  ngx_http_latency_conf_t *conf = child;
+  ngx_http_latency_main_conf_t *main_conf;
+
+  ngx_conf_merge_value(conf->receiver_enabled, prev->receiver_enabled, 0);
+  ngx_conf_merge_value(conf->status_page_enabled, prev->status_page_enabled, 0);
+  main_conf = ngx_http_conf_get_module_main_conf(cf, ngx_http_latency_stub_status_module);
+
+  if (!(main_conf->enabled) || main_conf->enabled == NGX_CONF_UNSET) {
+
+    if (conf->receiver_enabled) {
+      ngx_conf_log_error(NGX_LOG_ERR, cf, 0,
+                         "cannot enable latency recording for a location when latency storage is not enabled in the main conf.");
```

Comment:
I think we should also log that we are disabling it.

Reply:
I thought this was implied by "cannot enable".